spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java

```diff
@@ -159,16 +181,152 @@ public ListenableFuture<Job> writeDataToTable(
+   *     operation
    */
+  @Override
+  public ListenableFuture<WriteApiResponse> writeJsonStream(
+      String tableName, InputStream jsonInputStream, Schema schema) {
+    createTable(tableName, schema); // create table if it's not already created
+    return writeJsonStream(tableName, jsonInputStream);
+  }
+
+  private void createTable(
+      String tableName, Schema schema) { // create table if it's not already created
+    TableId tableId = TableId.of(datasetName, tableName);
+    Table table = bigQuery.getTable(TableId.of(datasetName, tableName));
+    if (table == null || !table.exists()) {
+      TableDefinition tableDefinition = StandardTableDefinition.of(schema);
+      TableInfo tableInfo = TableInfo.newBuilder(tableId, tableDefinition).build();
+      bigQuery.create(tableInfo);
+    }
+  }
+
+  /**
+   * This method uses BigQuery Storage Write API to write new line delimited JSON file to the
+   * specified table. The Table should already be created as BigQuery Storage Write API doesn't
+   * create it automatically.
+   *
+   * @param tableName name of the table to write to
+   * @param jsonInputStream input stream of the json file to be written
+   * @return {@link ListenableFuture} containing the WriteApiResponse indicating completion of
+   *     operation
+   */
+  @Override
+  public ListenableFuture<WriteApiResponse> writeJsonStream(
+      String tableName, InputStream jsonInputStream) {
+
+    SettableListenableFuture<WriteApiResponse> writeApiFutureResponse =
+        new SettableListenableFuture<>();
+    CompletableFuture<Void> completableFuture =
+        CompletableFuture.runAsync(
+            () -> {
+              try {
+                WriteApiResponse apiResponse = getWriteApiResponse(tableName, jsonInputStream);
+                writeApiFutureResponse.set(apiResponse);
+              } catch (DescriptorValidationException | IOException e) {
+                writeApiFutureResponse.setException(e);
+                logger.log(Level.WARNING, String.format("Error: %s\n", e.getMessage()));
+              } catch (InterruptedException e) {
+                writeApiFutureResponse.setException(e);
+                // Restore interrupted state in case of an InterruptedException
+                Thread.currentThread().interrupt();
+              }
+            });
+
+    // register success and failure callback
+    writeApiFutureResponse.addCallback(
+        (res) -> {
+          completableFuture.cancel(true); // interrupt the completableFuture
+        },
+        (res) -> {
+          completableFuture.cancel(true); // interrupt the completableFuture
+        });
+
+    return writeApiFutureResponse;
+  }
+
+  private WriteApiResponse getWriteApiResponse(String tableName, InputStream jsonInputStream)
+      throws DescriptorValidationException, IOException, InterruptedException {
+    WriteApiResponse apiResponse = new WriteApiResponse();
+    BigQueryWriteClient client = BigQueryWriteClient.create();
```

Comment:
This should be done as part of autoconfiguration, similar to [`BigQuery` object](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/e04f755f75c98619c11fde42106cff81e95f1540/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/bigquery/GcpBigQueryAutoConfiguration.java#L71) for two reasons -- 
1) to allow users to customize their own `BigQueryWriteClient` and provide it as part of user configuration.
2) to allow us to provide properties for customization that will automatically be used to configure the object in autoconfiguration.

Reply:
This makes sense, I have made a note of it and will update the code accordingly

Reply:
One more benefit is that it will make the code more testable -- if a `BigQueryWriteClient` object can come from the outside of the template, it can be mocked.

One thing I was not sure about -- is `BigQueryWriteClient` object meant to be reused between different write attempts (possibly to different tables)?

Reply:
Sure @elefeint , I have modified the code